### PR TITLE
Support jemalloc 5.x

### PR DESCRIPTION
**Pull request by botta633**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by wargio** [comment]
you forgot to install pyYAML which will allow you to regenerate the headers for newshell

**Comment by botta633** [comment]
> 

When I compiled the YAML files, the headers of jemalloc functions weren't generated. Does this imply a syntax error, or use of wrong command types..etc?

**Comment by wargio** [comment]
> > 
> 
> When I compiled the YAML files, the headers of jemalloc functions weren't generated. Does this imply a syntax error, or use of wrong command types..etc?

delete the build folder and run again `meson build && ninja -C build`

File: librz/core/cmd/cmd_linux_heap_jemalloc.c [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	{ \
+		if (core->rasm->bits == 64) { \
+			return fun##_64(core, ##__VA_ARGS__); \
+		} else { \
+			return fun##_32(core, ##__VA_ARGS__); \
+		} \
+	}
+
+RZ_IPI RzCmdStatus rz_cmd_jemalloc_print_narenas_handler(RzCore *core, int argc, const char **argv) {
+	call_handler(rz_cmd_jemalloc_print_narenas_handler, argc, argv);
```
Comment (wargio) [inline librz/core/cmd/cmd_linux_heap_jemalloc.c:14]: i do not see this handler being implemented.

**Comment by wargio** [comment]
Looks like your `yaml`file is not correctly set on the meson.build file
```
[1463/1768] Generating librz/core/cmd_descs/cmd_descs.[ch] with a custom command
FAILED: librz/core/cmd_descs/cmd_descs.c librz/core/cmd_descs/cmd_descs.h 
/usr/bin/python3 ../librz/core/cmd_descs/cmd_descs_generate.py --output-dir librz/core/cmd_descs --src-output-dir /home/runner/work/rizin/rizin/librz/core/cmd_descs ../librz/core/cmd_descs/cmd_analysis.yaml ../librz/core/cmd_descs/cmd_block.yaml ../librz/core/cmd_descs/cmd_cmp.yaml ../librz/core/cmd_descs/cmd_debug.yaml ../librz/core/cmd_descs/cmd_descs.yaml ../librz/core/cmd_descs/cmd_egg.yaml ../librz/core/cmd_descs/cmd_eval.yaml ../librz/core/cmd_descs/cmd_flag.yaml ../librz/core/cmd_descs/cmd_heap_glibc.yaml ../librz/core/cmd_descs/cmd_history.yaml ../librz/core/cmd_descs/cmd_info.yaml ../librz/core/cmd_descs/cmd_interpret.yaml ../librz/core/cmd_descs/cmd_meta.yaml ../librz/core/cmd_descs/cmd_open.yaml ../librz/core/cmd_descs/cmd_plugins.yaml ../librz/core/cmd_descs/cmd_print.yaml ../librz/core/cmd_descs/cmd_project.yaml ../librz/core/cmd_descs/cmd_quit.yaml ../librz/core/cmd_descs/cmd_remote.yaml ../librz/core/cmd_descs/cmd_resize.yaml ../librz/core/cmd_descs/cmd_seek.yaml ../librz/core/cmd_descs/cmd_shell.yaml ../librz/core/cmd_descs/cmd_system.yaml ../librz/core/cmd_descs/cmd_tasks.yaml ../librz/core/cmd_descs/cmd_type.yaml ../librz/core/cmd_descs/cmd_write.yaml ../librz/core/cmd_descs/cmd_yank.yaml ../librz/core/cmd_descs/cmd_zign.yaml
Command dmx referenced another YAML file (cmd_heap_jemalloc) that is not passed as arg to cmd_descs_generate.py.
```

File: librz/core/cmd/cmd_linux_heap_jemalloc.c [hunk 2]
```diff
@@ -0,0 +1,23 @@
+
```
Comment (XVilka) [inline librz/core/cmd/cmd_linux_heap_jemalloc.c:1]: Missing SPDX header.

File: librz/core/cmd_descs/cmd_heap_jemalloc.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_heap_jemalloc.yaml:1]: Missing SPDX header

File: librz/core/cmd_descs/cmd_heap_jemalloc.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  - name: dmxa
+    cname: cmd_jemalloc_print_nareans
+    summary: show all arenas created or print arena_t structure for given arena
+    type: RZ_CMD_DESC_TYPE_ARGV
+    args:
+      - name: arena_t
+        type: RZ_CMD_ARG_TYPE_STRING
+  - name: dmxb
+    cname: cmd_jemalloc_get_bins
+    summary: show all bins created for a gievn arena
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_heap_jemalloc.yaml:16]: `given`

File: librz/core/cmd_descs/cmd_heap_jemalloc.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#
+#
+#
+---
+name: cmd_heap_jemalloc
+commands:
+  - name: dmxa
+    cname: cmd_jemalloc_print_nareans
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_heap_jemalloc.yaml:8]: `narenas`

File: librz/core/cmd_descs/cmd_heap_jemalloc.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    summary: show all arenas created or print arena_t structure for given arena
+    type: RZ_CMD_DESC_TYPE_ARGV
+    args:
+      - name: arena_t
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_heap_jemalloc.yaml:12]: Just `arena`

File: librz/core/cmd_descs/cmd_heap_jemalloc.yaml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        type: RZ_CMD_ARG_TYPE_STRING
+  - name: dmxb
+    cname: cmd_jemalloc_get_bins
+    summary: show all bins created for a gievn arena
+    type: RZ_CMD_DESC_TYPE_ARGV
+    args:
+      - name: arena_t
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_heap_jemalloc.yaml:19]: Just `arena`

File: librz/core/cmd_descs/cmd_heap_jemalloc.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        type: RZ_CMD_ARG_TYPE_STRING
+  - name: dmxc
+    cname: cmd_jemalloc_get_chunks
+    summary: >
+      show all chunks created in all arenas, 
+      or show all chunks created for a given arena_t instance
+    type: RZ_CMD_DESC_TYPE_ARGV
+    args:
+      - name: arena_t
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_heap_jemalloc.yaml:28]: Just `arena`. Also in all cases it is probably optional argument.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/linux_heap_jemalloc.c [hunk 9]
```diff
@@ -58,21 +58,54 @@ static GHT GH(je_get_va_symbol)(RzCore *core, const char *path, const char *sym_
 	return vaddr;
 }
 
-static int GH(je_matched)(const char *ptr, const char *str) {
+static inline GHT GH(je_matched)(const char *ptr, const char *str) {
 	int ret = strncmp(ptr, str, strlen(str) - 1);
 	return !ret;
```
Comment (wargio) [inline librz/core/linux_heap_jemalloc.c]:
```suggestion
static bool GH(je_matched)(const char *ptr, const char *str) {
	return !strncmp(ptr, str, strlen(str) - 1);
```
why the `-1` ?

File: librz/core/linux_heap_jemalloc.c [hunk 10]
```diff
@@ -58,21 +58,54 @@ static GHT GH(je_get_va_symbol)(RzCore *core, const char *path, const char *sym_
...
 }
 #endif
+static GHT GH(get_main_arena_with_symbol)(RzCore *core, RzDebugMap *map) {
+    rz_return_val_if_fail(core && map, GHT_MAX);
+    GHT base_addr = map->addr;
+    rz_return_val_if_fail(base_addr != GHT_MAX, GHT_MAX);
```
Comment (wargio) [inline librz/core/linux_heap_jemalloc.c]:
```suggestion
    rz_return_val_if_fail(core && map && map->addr != GHT_MAX , GHT_MAX);
    GHT base_addr = map->addr;
```

File: librz/core/linux_heap_jemalloc.c (hunk 9 above)
Comment (botta633) [reply librz/core/linux_heap_jemalloc.c]: That is how it was  implemented initially, but ye what you say makes sense as by definition the null character is not counted

**Review by XVilka — CHANGES_REQUESTED** [review]
If you plan to finish this - please address the feedback, combine all commits into one or just few, rebase on top of the latest `dev`

**Comment by XVilka** [comment]
I am closing this since there was no response for almost half a year. If you are going to continue working on this - please open a new PR.

**Pull request closed without merging**